### ISSUE TITLE
no-boolean-literal-compare triggered on index type boolean value

### ISSUE THREAD

**Issue by Moeriki** [palantir/tslint#4602]
### Bug Report

- __5.14.0__:
- __3.2__:
- __Running TSLint via__: VSCode

#### TypeScript code being linted

```ts
let index: { [key: string]: boolean };
const isTrue = (key: string) => index[key] === true;
```

#### Actual behavior

* This expression is unnecessarily compared to a boolean. Just use it directly. (no-boolean-literal-compare).

#### Expected behavior

No warning. My reasoning is that the indexed value is either `undefined` or a boolean value. So I must check for `=== true`.

**Comment by Moeriki** [comment]
I currently write it like this, though it shouldn't be necessary.

```ts
const isTrue = (key: string) => key in index && index[key];
```

**Comment by JoshuaKGoldberg** [comment]
Hi @Moeriki! If your reasoning is that the indexed value is either `undefined` or a boolean value, you'll want to tell the type system that:

```ts
let index: { [key: string]: boolean | undefined };
``` 

In TypeScript's eyes, the type of `index[key]` will return whatever is declared in that index signature. Your original type is telling the type system that, for any key that will ever be retrieved from that object, it'll _only_ ever give `boolean` - never `undefined`.

**Comment by Moeriki** [comment]
Ok wasn't sure but that's sensible. Thanks for your feedback!
